include/zephyr/sys/hash_map_api.h

```diff
@@ -0,0 +1,194 @@
+ *
+ * @param key Key corresponding to @p value
+ * @param value Value corresponding to @p key
+ * @param cookie User-specified variable
+ */
+typedef void (*sys_hashmap_callback_t)(uint64_t key, uint64_t value, void *cookie);
+
+/**
+ * @brief Clear all entries contained in a @ref sys_hashmap
+ *
+ * @note If the values in a particular Hashmap are
+ *
+ * @param map Hashmap to clear
+ * @param cb Callback to call for each entry
+ * @param cookie User-specified variable
+ */
+typedef void (*sys_hashmap_clear_t)(struct sys_hashmap *map, sys_hashmap_callback_t cb,
+				    void *cookie);
+
+/**
+ * @brief Insert a new entry into a @ref sys_hashmap
+ *
+ * Insert a new @p key - @p value pair into @p map.
+ *
+ * @param map Hashmap to insert into
+ * @param key Key to associate with @p value
+ * @param value Value to associate with @p key
+ * @param old_value Location to store the value previously associated with @p key or `NULL`
+ * @retval 0 if @p value was inserted for an existing key, in which case @p old_value will contain
+ * the previous value
+ * @retval 1 if a new entry was inserted for the @p key - @p value pair
+ * @retval -ENOMEM if memory allocation failed
+ */
+typedef int (*sys_hashmap_insert_t)(struct sys_hashmap *map, uint64_t key, uint64_t value,
+				    uint64_t *old_value);
+
+/**
+ * @brief Remove an entry from a @ref sys_hashmap
+ *
+ * Erase the entry associated with key @p key, if one exists.
+ *
+ * @param map Hashmap to remove from
+ * @param key Key to remove from @p map
+ * @param value Location to store a potential value associated with @p key or `NULL`
+ *
+ * @retval true if @p map was modified as a result of this operation.
+ * @retval false if @p map does not contain a value associated with @p key.
+ */
+typedef bool (*sys_hashmap_remove_t)(struct sys_hashmap *map, uint64_t key, uint64_t *value);
+
+/**
+ * @brief Get a value from a @ref sys_hashmap
+ *
+ * Look-up the @ref uint64_t associated with @p key, if one exists.
+ *
+ * @param map Hashmap to search through
+ * @param key Key with which to search @p map
+ * @param value Location to store a potential value associated with @p key or `NULL`
+ *
+ * @retval true if @p map contains a value associated with @p key.
+ * @retval false if @p map does not contain a value associated with @p key.
+ */
+typedef bool (*sys_hashmap_get_t)(const struct sys_hashmap *map, uint64_t key, uint64_t *value);
+
+/**
+ * @brief Generic Hashmap API
+ *
+ * @param func Hash-generating function
+ * @param alloc Allocator with realloc-like semantics for internal Hashmap use
+ * @param iter Iterator constructor (in-place)
+ * @param next Forward-incrementer for iterator
+ * @param clear Clear the hash table, freeing all resources
+ * @param insert Insert a key-value pair into the Hashmap
+ * @param remove Remove a key-value pair from the Hashmap
+ * @param get Retrieve a the value associated with a given key from the Hashmap
+ */
+struct sys_hashmap_api {
+	sys_hash_func32_t func;
```

Comment:
nit: can we call this `hash_func` or `hash`?

Reply:
Sure - I have no major preference. Currently refactoring is not working for me in VS code though, so I'll do this later.